src/Type/Php/RegexShapeMatcher.php

```diff
@@ -0,0 +1,111 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PHPStan\Analyser\TypeSpecifierContext;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\Constant\ConstantArrayTypeBuilder;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\IntegerRangeType;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use function array_key_last;
+use function array_keys;
+use function is_string;
+use function preg_match;
+use function preg_replace;
+use const PREG_OFFSET_CAPTURE;
+use const PREG_UNMATCHED_AS_NULL;
+
+final class RegexShapeMatcher
+{
+
+	/**
+	 * @param int-mask<PREG_OFFSET_CAPTURE|PREG_UNMATCHED_AS_NULL>|null $flags
+	 */
+	public function matchType(string $regex, ?int $flags, TypeSpecifierContext $context): Type
+	{
+		if ($flags !== null) {
+			$trickFlags = PREG_UNMATCHED_AS_NULL | $flags;
+		} else {
+			$trickFlags = PREG_UNMATCHED_AS_NULL;
+		}
+
+		// add one capturing group to the end so all capture group keys
+		// are present in the $matches
+		// see https://3v4l.org/sOXbn, https://3v4l.org/3SdDM
+		$regex = preg_replace('~^(.)(.*)\K(\1\w*$)~', '|(?<phpstan_named_capture_group_last>)$3', $regex);
```

Comment:
I think that doesn't work for bracket style delimiters yet, like `'{regex}'`.
See https://www.php.net/manual/de/regexp.reference.delimiters.php.

TIL there can be leading whitespace before the delimiter: `'   /regex/'`

Reply:
Maybe this would work?

```suggestion
		$regex = preg_replace('~.[a-z\s]*$~i', '|(?<phpstan_named_capture_group_last>)$0', $regex);
```

I think it should work for all valid regexes. (And all invalid regexes should remain invalid.)

Note that I avoided to use `\w` for the modifiers at the end, because`\w` includes `_`, which can be used as delimiter.

Reply:
@gharlan could you also suggest a test-case for this change?

Reply:
I think these test cases would fail at the moment and I hope they are fixed when using my suggested regex:

```php
	if (preg_match('  /Price: (£|€)\d+/ i u', $s, $matches)) {
		assertType('array{0: string, 1?: string}', $matches);
	}
	assertType('array<string>', $matches);

	if (preg_match('(Price: (£|€))i', $s, $matches)) {
		assertType('array{0: string, 1?: string}', $matches);
	}
	assertType('array<string>', $matches);

	if (preg_match('_foo(.)\_i_i', $s, $matches)) {
		assertType('array{0: string, 1?: string}', $matches);
	}
	assertType('array<string>', $matches);
```

Reply:
it does work for the first 2, but not the last example

Reply:
(also interessting: with the new pattern, we get different errors per php version)

Reply:
Oh, I guess the `phpstan_named_capture_group_last` name is the problem there, because `_` is the delimiter in that regex. So it would be better to use `phpstanNamedCaptureGroupLast`.

Reply:
looks better now. PHP 7.2/7.3 errors remain.